grayscale.py

```diff
@@ -1,764 +1,549 @@
-                conf = fh.read().split(',')
-            for k in range(len(conf)):
-                if(conf[k] == "brightness"):
-                    brightnessSetting = int(conf[k+1])
-        except OSError:
+                _, _, conf = fh.read().partition("brightness,")
+                b = int(conf.split(',')[0])
+                # Set to the relevant brightness level
+                brightness = 0 if b==0 else 56 if b==1 else 127
+        except (OSError, ValueError):
             pass
-        # but with a set of contrast values spanning the entire range provided by the controller
-        brightnessVals = [0,56,127]
-        brightnessVal = brightnessVals[brightnessSetting]
-        # 0x81,<val>        Set Bank0 contrast value to <val>
-        self.post_frame_adj = [bytearray([0x81,brightnessVal>>5]), bytearray([0x81,brightnessVal]), bytearray([0x81,(brightnessVal << 1) + 1])]
+        # 0x81,<val>    Set Bank0 contrast value to <val>
+        self._postFrameAdj = array('O', [
+            bytearray([0x81,brightness>>5]),
+            bytearray([0x81,brightness]),
+            bytearray([0x81,(brightness << 1) + 1])])
 
-        # It's better to avoid using regular variables for thread sychronisation.
+        # It's safer to avoid using regular variables for shared thread data.
         # Instead, elements of an array/bytearray should be used.
-        # We're using a uint32 array here, as that should hopefully further ensure
+        # We're also using a uint32 array here, as thsis more likely to ensure
         # the atomicity of any element accesses.
-        self._state = array('I', [0,0,0,0xff])
-
-        self.pending_cmds = bytearray([0] * 8)
-
+        # [thread_state, buff_copy_gate, pending_cmd_gate, constrast_change]
+        self._state = array('I', [_THREAD_STOPPED,0,0,0xff])
+        # Buffer to funnel cmds to the thread
+        self._pendingCmds = bytearray([0] * 8)
+        # Set font, also supports: self.setFont('lib/font8x8.bin', 8, 8, 0)
         self.setFont('lib/font5x7.bin', 5, 7, 1)
-        #self.setFont('lib/font8x8.bin', 8, 8, 0)
-
-        self.lastUpdateEnd = 0
-        self.frameRate = 0
-        self.frameTimeMs = 0
 
-        self.fill(Grayscale.BLACK)
-        self._copy_buffers()
-        self.init_display()
-        self.state = _Grayscale_ThreadState_Starting
-        _thread.stack_size(2048)        # minimum stack size for RP2040 micropython port
-        _thread.start_new_thread(self._display_thread, ())
-
-        # Wait for the thread to successfully settle into a running state
-        while self._state[_Grayscale_StateIndex_State] != _Grayscale_ThreadState_Running:
-            idle()
-
-
-    # allow use of 'with'
+    # Allow use of 'with' for manaing the GPU state
     def __enter__(self):
         return self
     def __exit__(self, type, value, traceback):
-        self.stop()
+        self.stopGPU()
+    # Garbage collect the GPU thread properly
+    def __del__(self):
```

Comment:
I'll get rid of it. It's a singleton anyway so does nothing.

Reply:
It's gone now.

Reply:
I think GrayscaleFastTest used to soft reset back from running in the menu cleanly, and now doesn't. I think removing it may have removed something that did help in this scenario. I'm not going to add it back but just a note for if anyone wants to tackle this again.

Reply:
Further tests confirmed it did nothing. We managed to get the Thumby code updated to reset after every game.